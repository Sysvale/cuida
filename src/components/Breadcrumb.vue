<template>
	<nav>
		<ul class="breadcrumb">
			<li
				v-for="(item, index) in items"
				:key="`${index}-${item.name}-item`"
				role="presentation"
			>
				<cds-clickable
					v-if="index === 0"
					clickable
				>
					<router-link
						:to="routerPushTo(item)"
						class="breadcrumb__link"
					>
						<cds-icon
							v-if="items.length <= 2"
							height="16"
							width="16"
							name="arrow-left-outline"
						/>

						<cds-icon
							v-else
							height="15"
							width="15"
							name="home-outline"
						/>

						<cds-spacer margin-left="2">
							{{ item.label }}
						</cds-spacer>
					</router-link>
				</cds-clickable>

				<cds-spacer
					v-else
					class="breadcrumb__item"
					margin-right="1"
					margin-left="1"
				>
					<cds-icon
						height="16"
						width="16"
						name="caret-right-outline"
						class="breadcrumb__separator"
					/>

					<router-link
						:to="routerPushTo(item)"
						class="breadcrumb__link"
					>
						<cds-clickable :clickable="index !== items.length - 1">
							{{ item.label }}
						</cds-clickable>
					</router-link>
				</cds-spacer>
			</li>
		</ul>
	</nav>
</template>

<script>
import CdsIcon from './Icon.vue';
import CdsSpacer from './Spacer.vue';
import CdsClickable from './Clickable.vue';
import isEmpty from 'lodash.isempty';

export default {
	components: {
		CdsIcon,
		CdsSpacer,
		CdsClickable,
	},
	props: {
		/**
		 * Define a lista dos itens do Breadcrumb a serem
		 * mostrados. Os itens da lista devem ser
		 * objetos com path ou route, e com uma label.
		 */
		items: {
			type: Array,
			default: () => ([]),
			required: true,
		},
	},

	methods: {
		resolveRoute({ route, path }) {
			const to = isEmpty(route) ? path : route;
			return to instanceof String ? { path: to } : to;
		},

		routerPushTo(item) {
			return this.resolveRoute(item) ? this.resolveRoute(item).path : null;
		},
	},
};
</script>

<style lang="scss" scoped>
@import '../assets/sass/tokens.scss';

.breadcrumb {
	display: flex;
	align-items: center;
	padding: pa(0);

	&__item {
		@extend .breadcrumb;
	}

	&__link {
		@extend .breadcrumb;
		@include caption;
		color: $n-600;
		border-radius: $border-radius-lil;
		padding: pa(1);
		transition: $hover;

		&:hover,
		&:hover svg {
			background-color: $bn-50;
			color: $bn-600;
			transition: $hover;
		}
	}

	&__separator {
		color: $n-600;
	}
}
</style>
