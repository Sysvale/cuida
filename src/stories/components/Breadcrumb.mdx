import { Meta, Story, Props, Controls, Canvas, Preview } from '@storybook/blocks';
import Breadcrumb from '../../components/Breadcrumb.vue';
import * as BreadcrumbStories from './Breadcrumb.stories';

<Meta of={BreadcrumbStories} />

export const Template = (args) => ({
	components: { CdsBreadcrumb: Breadcrumb },
	setup() {
		return { args };
	},
	template: /*html*/ `
		<cds-breadcrumb
			v-bind="args"
		/>
	`,
});

# Breadcrumb

### Breadcrumbs são utilizados para indicar ao usuário o caminho percorrido entre as páginas de uma aplicação até se checar na página atual.

***

<br />

## Quando usar:

* Quando houver dois ou mais níveis na hierarquia de páginas;
* Quando você precisar indicar ao usuário em que página ele está e como ele pode refazer os passos que o levaram à página atual;

<br />

## Quando não usar:

* Em casos nos quais só houver uma página na estrutura de navegação.

<br />

## Obs.:

* Idealmente Breadcrumbs devem ser usados com <b>três</b> ou mais níveis de hierarquia. Entretanto, o Breadcrumb do
  Cuida pode ser usado com <b>dois</b> níveis pois nesse cenário ele se comporta como um botão de retorno para a página anterior.

<br />

## Preview

export const items = [
	{
		label: 'Relatórios',
		route: {
			path: '/reports',
			name: 'reports'
		},
	},
	{
		label: 'Relatórios individualizados',
		route: {
			path: '/individualized-reports',
			name: 'individualized'
		},
	},
	{
		label: 'Relatório de usuários',
		route: {
			path: '/users-reports',
			name: 'users'
		},
	},
];

export const items2 = [
	{
		label: 'Relatórios',
		route: {
			path: '/reports',
			name: 'reports'
		},
	},
	{
		label: 'Relatórios individualizados',
		route: {
			path: '/individualized-reports',
			name: 'individualized'
		},
	},
];

<br />

##### Breadcrumb de 3 ou mais níveis:

<Canvas>
  <Story of={BreadcrumbStories.Breadcrumb} />
</Canvas>

<Controls />

<br />

##### Breadcrumb de 2 níveis:

<Canvas>
  <Story of={BreadcrumbStories.TwoLevelBreadcrumb} />
</Canvas>
