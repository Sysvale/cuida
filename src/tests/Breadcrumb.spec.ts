import { describe, test, expect } from 'vitest';
import Breadcrumb from '../components/Breadcrumb.vue';
import { shallowMount } from '@vue/test-utils';

const items = [
	{
		label: 'Relat칩rios',
		route: {
			path: '/reports',
			name: 'reports'
		},
	},
	{
		label: 'Relat칩rios individualizados',
		route: {
			path: '/individualized-reports',
			name: 'individualized'
		},
	},
	{
		label: 'Relat칩rio de usu치rios',
		route: {
			path: '/users-reports',
			name: 'users'
		},
	},
];

describe('Breadcrumb snapshot test', () => {
	test('renders correctly', () => {
		const wrapper = shallowMount(Breadcrumb, {
			global: {
				stubs: {
					'router-link': true,
				},
			},
			props: {
				items: items,
			},
		});

		expect(wrapper.html()).toMatchSnapshot();
	});
});
