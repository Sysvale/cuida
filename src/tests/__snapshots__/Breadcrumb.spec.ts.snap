// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Breadcrumb snapshot test > renders correctly 1`] = `
"<nav data-v-1a0f46d9="">
  <ul data-v-1a0f46d9="" class="breadcrumb">
    <li data-v-1a0f46d9="" role="presentation">
      <cds-clickable-stub data-v-1a0f46d9="" clickable="true" fluid="false"></cds-clickable-stub>
    </li>
    <li data-v-1a0f46d9="" role="presentation">
      <cds-spacer-stub data-v-1a0f46d9="" padding="0" paddingtop="0" paddingright="0" paddingbottom="0" paddingleft="0" margin="0" margintop="0" marginright="1" marginbottom="0" marginleft="1" class="breadcrumb__item"></cds-spacer-stub>
    </li>
    <li data-v-1a0f46d9="" role="presentation">
      <cds-spacer-stub data-v-1a0f46d9="" padding="0" paddingtop="0" paddingright="0" paddingbottom="0" paddingleft="0" margin="0" margintop="0" marginright="1" marginbottom="0" marginleft="1" class="breadcrumb__item"></cds-spacer-stub>
    </li>
  </ul>
</nav>"
`;
